http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/http2/Http2MultiplexedChannelPool.java

```diff
@@ -46,151 +58,333 @@
+                            + "streams was not initialized.", parentChannel);
+            Validate.isTrue(maxStreams > 0, "Maximum streams were negative on channel (%s).", parentChannel);
+
+            MultiplexedChannelRecord multiplexedChannel = new MultiplexedChannelRecord(parentChannel, maxStreams);
+            parentChannel.attr(MULTIPLEXED_CHANNEL).set(multiplexedChannel);
+
+            Promise<Channel> streamPromise = parentChannel.eventLoop().newPromise();
+
+            if (!acquireStreamOnInitializedConnection(multiplexedChannel, streamPromise)) {
+                failAndCloseParent(promise, parentChannel,
+                                   new IOException("Connection was closed while creating a new stream."));
+                return;
             }
+
+            streamPromise.addListener(f -> {
+                if (!streamPromise.isSuccess()) {
+                    promise.setFailure(streamPromise.cause());
+                    return;
+                }
+
+                Channel stream = streamPromise.getNow();
+                cacheConnectionForFutureStreams(stream, multiplexedChannel, promise);
+            });
+        } catch (Throwable e) {
+            failAndCloseParent(promise, parentChannel, e);
+        }
+    }
+
+    private void cacheConnectionForFutureStreams(Channel stream,
+                                                 MultiplexedChannelRecord multiplexedChannel,
+                                                 Promise<Channel> promise) {
+        Channel parentChannel = stream.parent();
+
+        // Before we cache the connection, make sure that exceptions on the connection will remove it from the cache.
+        parentChannel.pipeline().addLast(new ReleaseOnExceptionHandler());
+
+        connections.add(multiplexedChannel);
+
+        if (closed.get()) {
+            // Whoops, we were closed while we were setting up. Make sure everything here is cleaned up properly.
+            failAndCloseParent(promise, parentChannel,
+                               new IOException("Connection pool was closed while creating a new stream."));
+            return;
         }
-        connections.remove(record);
+
+        promise.setSuccess(stream);
+    }
+
+    private Void failAndCloseParent(Promise<Channel> promise, Channel parentChannel, Throwable exception) {
+        promise.setFailure(exception);
+        closeAndReleaseParent(parentChannel);
+        return null;
+    }
+
+    /**
+     * Acquire a stream on a connection that has already been initialized. This will return false if the connection cannot have
+     * any more streams allocated, and true if the stream can be allocated.
+     *
+     * This will NEVER complete the provided future when the return value is false. This will ALWAYS complete the provided
+     * future when the return value is true.
+     */
+    private boolean acquireStreamOnInitializedConnection(MultiplexedChannelRecord channelRecord, Promise<Channel> promise) {
```

Comment:
The previous check doesn't work in the world that the Http2MultiplexedChannelPool is multithreaded, because checking whether a stream is available and acquiring that stream isn't an atomic operation. Do you have a proposed alternative?